### PR TITLE
Initial support for Single Page Story Ads on DoubleClick

### PR DESCRIPTION
**Pull request by glevitzky**
In anticipation of Single Page Story Ads (SPSA) in DoubleClick, this PR does the following:

- If the slot has been marked by the runtime with the `amp-story` attribute, the slot should be aware that it is in an SPSA situation.
- If in SPSA situation, add `spsa=${viewport_width}x${viewport_height}` to ad request.
- If in SPSA situation and invalid AMP creative is served, throw error.
- If in SPSA situation, look for the call to action (CTA) enum and the outlink in the metadata object.
- If both properties are found on the metadata object, add them to appropriate data attributes on the slot, otherwise throw error.

### CODE REVIEW

File: extensions/amp-a4a/0.1/amp-a4a.js [hunk 1]
```diff
@@ -806,8 +809,9 @@ export class AmpA4A extends AMP.BaseElement {
           return creativeMetaDataDef;
         })
         .catch(error => {
-          switch (error) {
+          switch (error.message || error) {
             case NETWORK_FAILURE: return null;
+            case INVALID_SPSA_RESPONSE:
```
[the remark this answers is not in the document]
Comment (glevitzky) [reply extensions/amp-a4a/0.1/amp-a4a.js]: Fixed.

File: extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js [hunk 2]
```diff
@@ -1339,6 +1349,34 @@ export class AmpAdNetworkDoubleclickImpl extends AmpA4A {
   getA4aAnalyticsConfig() {
     return getCsiAmpAnalyticsConfig();
   }
+
+  /** @override */
+  maybeValidateAmpCreative(bytes, headers) {
```
[the remark this answers is not in the document]
Comment (glevitzky) [reply extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js]: Moved things over to amp-a4a.

File: extensions/amp-a4a/0.1/amp-a4a.js [hunk 3]
```diff
@@ -449,6 +460,8 @@ export class AmpA4A extends AMP.BaseElement {
       this.a4aAnalyticsElement_ = insertAnalyticsElement(
           this.element, this.a4aAnalyticsConfig_, true /* loadAnalytics */);
     }
+
+    this.isSinglePageStoryAd = 'ampStory' in this.element.dataset;
```
Comment (calebcordry) [inline extensions/amp-a4a/0.1/amp-a4a.js]: Right now this is not `data-` prefixed so I don't think this check will work. `this.element.hasAttribute('amp-story')`

**Comment by calebcordry** [comment]
I will make a manual test today and cc/ everyone so that we can make sure everything  is behaving as expected.

File: extensions/amp-a4a/0.1/amp-a4a.js (hunk 3 above)
Comment (glevitzky) [reply extensions/amp-a4a/0.1/amp-a4a.js]: Fixed.

**Comment by glevitzky** [comment]
Verified e2e, and things LGTM. Feel free to merge.

**Pull request merged**